dm/master/server.go

```diff
@@ -1559,3 +1559,141 @@ func (s *Server) getSourceRespsAfterOperation(ctx context.Context, taskName stri
 	wg.Wait()
 	return sortCommonWorkerResults(sourceRespCh)
 }
+
+func (s *Server) listMemberMaster(ctx context.Context) (*pb.Members_Master, error) {
+
+	resp := &pb.Members_Master{
+		Master: &pb.ListMasterMember{},
+	}
+
+	memberList, err := s.etcdClient.MemberList(ctx)
+	if err != nil {
+		resp.Master.Msg = errors.ErrorStack(err)
+		return resp, nil
+	}
+	etcdMembers := memberList.Members
+	masters := make([]*pb.MasterInfo, 0, len(etcdMembers))
+	client := http.Client{
+		Timeout: 1 * time.Second,
+	}
+
+	for _, etcdMember := range etcdMembers {
+		alive := true
+		_, err := client.Get(etcdMember.ClientURLs[0] + "/health")
+		if err != nil {
+			alive = false
+		}
+
+		masters = append(masters, &pb.MasterInfo{
+			Name:       etcdMember.Name,
+			MemberID:   etcdMember.ID,
+			Alive:      alive,
+			ClientURLs: etcdMember.ClientURLs,
+			PeerURLs:   etcdMember.PeerURLs,
+		})
+	}
+
+	resp.Master.Masters = masters
+	return resp, nil
+}
+
+func (s *Server) listMemberWorker(ctx context.Context) (*pb.Members_Worker, error) {
+	resp := &pb.Members_Worker{
+		Worker: &pb.ListWorkerMember{},
+	}
+
+	workerAgents := s.scheduler.GetWorker()
+	workers := make([]*pb.WorkerInfo, 0, len(workerAgents))
+
+	for _, workerAgent := range workerAgents {
+		workers = append(workers, &pb.WorkerInfo{
+			Name:   workerAgent.BaseInfo().Name,
+			Addr:   workerAgent.BaseInfo().Addr,
+			Stage:  string(workerAgent.Stage()),
+			Source: workerAgent.Bound().Source,
+		})
+	}
+
+	resp.Worker.Workers = workers
+	return resp, nil
+}
+
+func (s *Server) listMemberLeader(ctx context.Context) (*pb.Members_Leader, error) {
+	resp := &pb.Members_Leader{
+		Leader: &pb.ListLeaderMember{},
+	}
+
+	_, name, addr, err := s.election.LeaderInfo(ctx)
+	if err != nil {
+		resp.Leader.Msg = errors.ErrorStack(err)
+		return resp, nil
+	}
+
+	client := http.Client{
+		Timeout: 1 * time.Second,
+	}
+
+	alive := true
+	_, err = client.Get("http://" + addr + "/health")
```

Comment:
in which case the leader is not alive?

Reply:
Maybe when we enter this function the leader change and immediately down? Alright I will remove it. 😂